### ISSUE TITLE
Custom loaders.

### ISSUE THREAD

**Issue by Lattyware** [Lattyware/unrpa#15]
It appears some Ren'Py games are starting to ship with custom loader scripts for a non-standard variant of RPA archives. `unrpa` currently can't deal with these archives.

The archives that do this seen in the wild seem to be identifiable as they begin with a `ZiX-12B` header, not the expected `RPA-3.0`/`RPA-2.0`. This appears to be an in-house obfuscation technique.

The route to decoding these files is to use [`uncompyle6`](https://pypi.org/project/uncompyle6/3.2.3/) to turn 'loader.pyo' from the game the archive comes from into readable code. This should allow you to modify `unrpa` to load the archive. It appears to use a compiled cython module called `_string` to perform parts of the process.

It appears the system is to use a hard-coded hey in the loader. Ideally, we could identify this type of archive by the header and offer additional tooling to extract that key, alongside an option to manually set the key as an argument.

(This is the root cause of #13).

Edit: There is [a script](https://github.com/Lattyware/unrpa/issues/15#issuecomment-419853604) to make extracting these possible, but proper support isn't here yet. See below for details on how to extract an archive of this type now.

Edit: For transparency, I will note I worked out who the developer was who created this technique, and had their name listed here previously. At their request, I have removed a direct reference to them from this post, as it's not really relevant. The partial support for the format and the documentation of the effort here will remain up, however. I am still happy to accept pull requests to solve this issue properly and add full support to unrpa.

**Comment by Lattyware** [comment]
You can extract these files with the following process:

- Take the `_string.pyd`/`_string.so` module for your platform from the `lib` sub-directory for your platform (it's in different places depending on the platform).
- Take `loader.pyo` from `renpy`
- Use `uncompyle6` to get `loader.py` from `loader.pyo`.
- Look for a line setting `verificationcode` (e.g: `verificationcode = _string.sha1('1a7ee58d44d3c6dc943600c2b0c7f13670a6f5a0')`)
- Pull the first line from the archive (e.g: `ZiX-12B 000000000a849dd5 6e641880`).
- Run python 2.x in the directory containing the `_string` module obtained earlier. `import _string` and then run the `verificationcode` line - the result of this is your `key`.
- Now run `_string.offset()` with the argument as the last section of the first line of the archive (e.g: `_string.offset("6e641880")`). The result of this is your `offset`.

Now modify `unrpa` to replace the `offset` and `key` under `elif self.version == 3:` with the values you just obtained. If you force the version to `3`, this should now successfully decode the archive.

Clearly, this is a massive pain. It could be made a little nicer by having command line arguments to override `key`/`offset`, and further by reverse-engineering the two methods from the `_string` module, which is presumably a `cython` module.

**Comment by Lattyware** [comment]
There appears to still be an issue - while this produces a directory structure and files that look reasonable, the actual files produced don't appear to be correct. There is presumably still some other obfuscation going on.

**Comment by Lattyware** [comment]
The latest version has some extra handling to point the user here if they try and extract an archive of this type.

I have also made [a secondary script](https://gist.github.com/Lattyware/c7ae85998f62a5d1a985de6bfc662048) that automates the above process - it will be somewhat fragile and still relies on the original `_string` module, meaning it isn't ideal. Reverse engineering that module will still be needed for proper support, but this should make it easier until then.

**Comment by Lattyware** [comment]
The file will be inside the `lib` folder in the `renpy` folder of the game you are trying to extract from. The exact path will depend on your platform (Windows, Linux, Mac).

**Comment by omegalink12** [comment]
I managed to track down a copy of `_string.pyd` (MD5 BCD019154309731EB1780546E2E82155) and reverse it. I now know more about cython internals than I ever wanted to. I made a python version of the module that should be easy enough to integrate. I've tested it on random input but not a complete archive. Looking at other games by the same company, they have different loader versions you could also support.

```py
import struct

def sha1(code):
	a=int(filter(str.isdigit,code))+102464652121606009
	b=round(a**(1./3))/23*109
	return int(b)

def offset(offset):
	a=offset[7:5:-1]
	b=offset[:3]
	c=offset[5:2:-1]
	return int(a+b+c,16)

def run(s,key):
	keys=(3621826839565189698,8167163782024462963,5643161164948769306,4940859562182903807,2672489546482320731,8917212212349173728,7093854916990953299)
	out=''
	for i in range(0,len(s),8):
		enc=struct.unpack("<Q",s[i:i+8])[0]
		dec=keys[i%7]^key^enc
		out=out+struct.pack("<Q",dec)
	return out
```

**Comment by Lattyware** [comment]
Nice work! I took a look at trying to reverse engineer it myself and it looked like a massive pain in the ass, so congrats on getting through that. As soon as I have the chance I'll take a look at integrating this into `unrpa` properly, which should be trivial enough given the pure-python implementation you have provided.

**Comment by Lattyware** [comment]
Resolved as of 2.0.0.

**Comment by omegalink12** [comment]
Minor Note: obfuscated_amount is also loader dependent.

**Comment by omegalink12** [comment]
In general, we need to take care with post processing as it only applies to some rpa archives

**Comment by Lattyware** [comment]
The default `postprocessing()` implementation is just a pass-through that does nothing - only the ZiX-12B implementation does anything there.

I see how `obfuscated_amount` could be varied. I'll write up a fix that is dynamic over that. If you have any other examples of the format, I'd love to have some more test cases.

**Comment by omegalink12** [comment]
The problem with the ZiX-12B implementation is that it applies postprocessing to all zix archives. However, the loader only applies it to specific ones. The list of said archives is also loader specific. See other VNs by the same company for example.

**Comment by Lattyware** [comment]
Oh, I see. That's something I didn't even think to look for. I'll fix that along with the other change and push a new version when I get a chance. Let me know if there is anything else you notice, and thanks for all the help getting this one implemented.

**Comment by Lattyware** [comment]
Current concerns should be fixed as of 7eecd64ea912b5a7187ffd4e353ae4bd7dc98647. Let me know if anything else comes up.
